#[axum::debug_hander] causes compilation error

Hi, thanks for your work on this! I'm trying it out in a project.

One issue I ran into was using `[axum::debug_handler]` to troubleshoot something, then being left with another error after I fixed the initial problem. The error seems to occur if it's used on any handler that returns `impl IntoCodecResponse` e.g.:

```rust
error[E0277]: the trait bound `impl IntoCodecResponse: IntoResponse` is not satisfied
  --> examples/basic/src/main.rs:19:18
   |
19 | async fn me() -> impl IntoCodecResponse {
   |                  ^^^^ the trait `IntoResponse` is not implemented for `impl IntoCodecResponse`
   |
   = help: the following other types implement trait `IntoResponse`:
             &'static [u8; N]
             &'static [u8]
             &'static str
             ()
             (R,)
             (Response<()>, R)
             (Response<()>, T1, R)
             (Response<()>, T1, T2, R)
           and 122 others
note: required by a bound in `__axum_macros_check_me_into_response::{closure#0}::check`
  --> examples/basic/src/main.rs:19:18
   |
19 | async fn me() -> impl IntoCodecResponse {
   |                  ^^^^ required by this bound in `check`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `basic` (bin "basic") due to 1 previous error

```

It took me a while to realize that I just needed to remove `debug_handler` to fix the error.

--- a/README.md
+++ b/README.md
@@ -1,7 +1,7 @@
 # Axum Codec
 
-[![https://img.shields.io/crates/v/axum-codec](https://img.shields.io/crates/v/axum-codec)](https://crates.io/crates/axum-codec)
-[![https://img.shields.io/docsrs/axum-codec](https://img.shields.io/docsrs/axum-codec)](https://docs.rs/axum-codec/latest/axum_codec/)
+[![<https://img.shields.io/crates/v/axum-codec>](https://img.shields.io/crates/v/axum-codec)](https://crates.io/crates/axum-codec)
+[![<https://img.shields.io/docsrs/axum-codec>](https://img.shields.io/docsrs/axum-codec)](https://docs.rs/axum-codec/latest/axum_codec/)
 [![ci status](https://github.com/matteopolak/axum-codec/workflows/ci/badge.svg)](https://github.com/matteopolak/axum-codec/actions)
 
 A body extractor for the [Axum](https://github.com/tokio-rs/axum) web framework.
@@ -23,6 +23,7 @@ A body extractor for the [Axum](https://github.com/tokio-rs/axum) web framework.
 - [ ] Add benchmarks?
 
 Here's a quick example that can do the following:
+
 - Decode a `User` from the request body in any of the supported formats.
 - Encode a `Greeting` to the response body in any of the supported formats.
 
@@ -98,7 +99,7 @@ async fn main() {
 # Feature flags
 
 - `macros`: Enables the `axum_codec::apply` attribute macro.
-- `json`*: Enables [`JSON`](https://github.com/serde-rs/json) support.
+- `json`\*: Enables [`JSON`](https://github.com/serde-rs/json) support.
 - `msgpack`: Enables [`MessagePack`](https://github.com/3Hren/msgpack-rust) support.
 - `bincode`: Enables [`Bincode`](https://github.com/bincode-org/bincode) support.
 - `bitcode`: Enables [`Bitcode`](https://github.com/SoftbearStudios/bitcode) support.
@@ -110,7 +111,11 @@ async fn main() {
 
 * Enabled by default.
 
+## A note about `#[axum::debug_handler]`
+
+Since `axum-codec` uses its own `IntoCodecResponse` trait for encoding responses, it is not compatible with `#[axum::debug_handler]`. However, a new `#[axum_codec::debug_handler]` (and `#[axum_codec::debug_middleware]`) macro
+is provided as a drop-in replacement.
+
 ## License
 
 Dual-licensed under MIT or Apache License v2.0.
-

--- a/examples/aide-validator/src/main.rs
+++ b/examples/aide-validator/src/main.rs
@@ -14,6 +14,7 @@ struct User {
 	age: u8,
 }
 
+#[axum_codec::debug_handler]
 async fn me() -> impl IntoCodecResponse {
 	Codec(User {
 		name: "Alice".into(),

--- a/macros/Cargo.toml
+++ b/macros/Cargo.toml
@@ -16,6 +16,9 @@ quote = "1"
 syn = { version = "2", optional = true }
 
 [features]
+default = ["debug"]
+
+debug = []
 bincode = ["dep:syn"]
 bitcode = ["dep:syn"]
 serde = ["dep:syn"]

--- a/macros/src/apply.rs
+++ b/macros/src/apply.rs
@@ -1,0 +1,202 @@
+use proc_macro2::TokenStream;
+use quote::{quote, ToTokens};
+use syn::{
+	parse::{Parse, ParseStream},
+	punctuated::Punctuated,
+	spanned::Spanned,
+	Meta, Path, Token,
+};
+
+struct Args {
+	encode: bool,
+	decode: bool,
+	crate_name: Path,
+}
+
+impl Parse for Args {
+	fn parse(input: ParseStream) -> syn::Result<Self> {
+		let options = Punctuated::<Meta, Token![,]>::parse_terminated(input)?;
+
+		let mut encode = false;
+		let mut decode = false;
+		let mut crate_name = syn::parse_str("axum_codec").expect("failed to parse crate name");
+
+		for meta in options {
+			match meta {
+				Meta::List(list) => {
+					return Err(syn::Error::new(
+						list.span(),
+						"expected `encode`, `decode`, or `crate`",
+					))
+				}
+				Meta::Path(path) => {
+					let ident = path.get_ident().map(|ident| ident.to_string());
+					match ident.as_deref() {
+						Some("encode") if encode => {
+							return Err(syn::Error::new(
+								path.span(),
+								"option `encode` is already enabled",
+							))
+						}
+						Some("decode") if decode => {
+							return Err(syn::Error::new(
+								path.span(),
+								"option `decode` is already enabled",
+							))
+						}
+						Some("encode") => encode = true,
+						Some("decode") => decode = true,
+						Some(other) => {
+							return Err(syn::Error::new(
+								path.span(),
+								format!("unknown option `{other}`, expected `encode` or `decode`"),
+							))
+						}
+						None => {
+							return Err(syn::Error::new(
+								path.span(),
+								"expected `encode` or `decode`",
+							))
+						}
+					}
+				}
+				Meta::NameValue(name_value) => {
+					if !name_value.path.is_ident("crate") {
+						return Err(syn::Error::new(name_value.path.span(), "expected `crate`"));
+					}
+
+					let path = match name_value.value {
+						syn::Expr::Lit(ref lit) => match &lit.lit {
+							syn::Lit::Str(path) => path,
+							_ => return Err(syn::Error::new(lit.span(), "expected a string")),
+						},
+						_ => {
+							return Err(syn::Error::new(
+								name_value.value.span(),
+								"expected a literal string",
+							))
+						}
+					};
+
+					let mut path = syn::parse_str::<Path>(&path.value()).expect("failed to parse path");
+
+					path.leading_colon = if path.is_ident("crate") {
+						None
+					} else {
+						Some(Token![::](name_value.value.span()))
+					};
+
+					crate_name = path;
+				}
+			}
+		}
+
+		if !encode && !decode {
+			return Err(syn::Error::new(
+				input.span(),
+				"at least one of `encode` or `decode` must be enabled",
+			));
+		}
+
+		Ok(Self {
+			encode,
+			decode,
+			crate_name,
+		})
+	}
+}
+
+pub fn apply(
+	attr: proc_macro::TokenStream,
+	input: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+	let args = syn::parse_macro_input!(attr as Args);
+
+	let crate_name = &args.crate_name;
+	let mut tokens = TokenStream::default();
+
+	#[cfg(feature = "serde")]
+	{
+		if args.encode {
+			tokens.extend(quote! {
+				#[derive(#crate_name::__private::serde::Serialize)]
+			});
+		}
+
+		if args.decode {
+			tokens.extend(quote! {
+				#[derive(#crate_name::__private::serde::Deserialize)]
+			});
+		}
+
+		let crate_ = format!("{}::__private::serde", crate_name.to_token_stream());
+
+		tokens.extend(quote! {
+			#[serde(crate = #crate_)]
+		});
+	}
+
+	#[cfg(feature = "bincode")]
+	{
+		if args.encode {
+			tokens.extend(quote! {
+				#[derive(#crate_name::__private::bincode::Encode)]
+			});
+		}
+
+		if args.decode {
+			tokens.extend(quote! {
+				#[derive(#crate_name::__private::bincode::Decode)]
+			});
+		}
+
+		let crate_ = format!("{}::__private::bincode", crate_name.to_token_stream());
+
+		tokens.extend(quote! {
+			#[bincode(crate = #crate_)]
+		});
+	}
+
+	#[cfg(feature = "bitcode")]
+	{
+		if args.encode {
+			tokens.extend(quote! {
+				#[derive(#crate_name::__private::bitcode::Encode)]
+			});
+		}
+
+		if args.decode {
+			tokens.extend(quote! {
+				#[derive(#crate_name::__private::bitcode::Decode)]
+			});
+		}
+
+		let crate_ = format!("{}::__private::bitcode", crate_name.to_token_stream());
+
+		tokens.extend(quote! {
+			#[bitcode(crate = #crate_)]
+		});
+	}
+
+	#[cfg(feature = "aide")]
+	{
+		let crate_ = format!("{}::__private::schemars", crate_name.to_token_stream());
+
+		tokens.extend(quote! {
+			#[derive(#crate_name::__private::schemars::JsonSchema)]
+			#[schemars(crate = #crate_)]
+		});
+	}
+
+	// TODO: Implement #[validate(crate = "...")]
+	// For now, use the real crate name so the error is nicer.
+	#[cfg(feature = "validator")]
+	if args.decode {
+		tokens.extend(quote! {
+			#[derive(validator::Validate)]
+		});
+	}
+
+	tokens.extend(TokenStream::from(input));
+	tokens.into()
+}

--- a/macros/src/attr_parsing.rs
+++ b/macros/src/attr_parsing.rs
@@ -1,0 +1,60 @@
+// This is copied from Axum under the following license:
+//
+// Copyright 2021 Axum Contributors
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use quote::ToTokens;
+use syn::{
+	parse::{Parse, ParseStream},
+	Token,
+};
+
+pub(crate) fn parse_assignment_attribute<K, T>(
+	input: ParseStream,
+	out: &mut Option<(K, T)>,
+) -> syn::Result<()>
+where
+	K: Parse + ToTokens,
+	T: Parse,
+{
+	let kw = input.parse()?;
+	input.parse::<Token![=]>()?;
+	let inner = input.parse()?;
+
+	if out.is_some() {
+		let kw_name = std::any::type_name::<K>().split("::").last().unwrap();
+		let msg = format!("`{kw_name}` specified more than once");
+		return Err(syn::Error::new_spanned(kw, msg));
+	}
+
+	*out = Some((kw, inner));
+
+	Ok(())
+}
+
+pub(crate) fn second<T, K>(tuple: (T, K)) -> K {
+	tuple.1
+}

--- a/macros/src/debug_handler.rs
+++ b/macros/src/debug_handler.rs
@@ -1,0 +1,911 @@
+// This is copied from Axum under the following license:
+//
+// Copyright 2021 Axum Contributors
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::{collections::HashSet, fmt};
+
+use proc_macro2::{Ident, Span, TokenStream};
+use quote::{format_ident, quote, quote_spanned};
+use syn::{parse::Parse, spanned::Spanned, FnArg, ItemFn, ReturnType, Token, Type};
+
+use crate::{
+	attr_parsing::{parse_assignment_attribute, second},
+	with_position::{Position, WithPosition},
+};
+
+pub(crate) fn expand(attr: Attrs, item_fn: ItemFn, kind: FunctionKind) -> TokenStream {
+	let Attrs { state_ty } = attr;
+
+	let mut state_ty = state_ty.map(second);
+
+	let check_extractor_count = check_extractor_count(&item_fn, kind);
+	let check_path_extractor = check_path_extractor(&item_fn, kind);
+	let check_output_tuples = check_output_tuples(&item_fn);
+	let check_output_impls_into_response = if check_output_tuples.is_empty() {
+		check_output_impls_into_response(&item_fn)
+	} else {
+		check_output_tuples
+	};
+
+	// If the function is generic, we can't reliably check its inputs or whether the
+	// future it returns is `Send`. Skip those checks to avoid unhelpful additional
+	// compiler errors.
+	let check_inputs_and_future_send = if item_fn.sig.generics.params.is_empty() {
+		let mut err = None;
+
+		if state_ty.is_none() {
+			let state_types_from_args = state_types_from_args(&item_fn);
+
+			#[allow(clippy::comparison_chain)]
+			if state_types_from_args.len() == 1 {
+				state_ty = state_types_from_args.into_iter().next();
+			} else if state_types_from_args.len() > 1 {
+				err = Some(
+					syn::Error::new(
+						Span::call_site(),
+						format!(
+							"can't infer state type, please add set it explicitly, as in \
+							 `#[axum_codec::debug_{kind}(state = MyStateType)]`"
+						),
+					)
+					.into_compile_error(),
+				);
+			}
+		}
+
+		err.unwrap_or_else(|| {
+			let state_ty = state_ty.unwrap_or_else(|| syn::parse_quote!(()));
+
+			let check_future_send = check_future_send(&item_fn, kind);
+
+			if let Some(check_input_order) = check_input_order(&item_fn, kind) {
+				quote! {
+						#check_input_order
+						#check_future_send
+				}
+			} else {
+				let check_inputs_impls_from_request =
+					check_inputs_impls_from_request(&item_fn, state_ty, kind);
+
+				quote! {
+						#check_inputs_impls_from_request
+						#check_future_send
+				}
+			}
+		})
+	} else {
+		syn::Error::new_spanned(
+			&item_fn.sig.generics,
+			format!("`#[axum_codec::debug_{kind}]` doesn't support generic functions"),
+		)
+		.into_compile_error()
+	};
+
+	let middleware_takes_next_as_last_arg =
+		matches!(kind, FunctionKind::Middleware).then(|| next_is_last_input(&item_fn));
+
+	quote! {
+			#item_fn
+			#check_extractor_count
+			#check_path_extractor
+			#check_output_impls_into_response
+			#check_inputs_and_future_send
+			#middleware_takes_next_as_last_arg
+	}
+}
+
+#[derive(Clone, Copy)]
+pub(crate) enum FunctionKind {
+	Handler,
+	Middleware,
+}
+
+impl fmt::Display for FunctionKind {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		match self {
+			FunctionKind::Handler => f.write_str("handler"),
+			FunctionKind::Middleware => f.write_str("middleware"),
+		}
+	}
+}
+
+impl FunctionKind {
+	fn name_uppercase_plural(&self) -> &'static str {
+		match self {
+			FunctionKind::Handler => "Handlers",
+			FunctionKind::Middleware => "Middleware",
+		}
+	}
+}
+
+mod kw {
+	syn::custom_keyword!(body);
+	syn::custom_keyword!(state);
+}
+
+pub(crate) struct Attrs {
+	state_ty: Option<(kw::state, Type)>,
+}
+
+impl Parse for Attrs {
+	fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+		let mut state_ty = None;
+
+		while !input.is_empty() {
+			let lh = input.lookahead1();
+			if lh.peek(kw::state) {
+				parse_assignment_attribute(input, &mut state_ty)?;
+			} else {
+				return Err(lh.error());
+			}
+
+			let _ = input.parse::<Token![,]>();
+		}
+
+		Ok(Self { state_ty })
+	}
+}
+
+fn check_extractor_count(item_fn: &ItemFn, kind: FunctionKind) -> Option<TokenStream> {
+	let max_extractors = 16;
+	let inputs = item_fn
+		.sig
+		.inputs
+		.iter()
+		.filter(|arg| skip_next_arg(arg, kind))
+		.count();
+	if inputs <= max_extractors {
+		None
+	} else {
+		let error_message = format!(
+			"{} cannot take more than {max_extractors} arguments. Use `(a, b): (ExtractorA, \
+			 ExtractorA)` to further nest extractors",
+			kind.name_uppercase_plural(),
+		);
+		let error = syn::Error::new_spanned(&item_fn.sig.inputs, error_message).to_compile_error();
+		Some(error)
+	}
+}
+
+fn extractor_idents(
+	item_fn: &ItemFn,
+	kind: FunctionKind,
+) -> impl Iterator<Item = (usize, &syn::FnArg, &syn::Ident)> {
+	item_fn
+		.sig
+		.inputs
+		.iter()
+		.filter(move |arg| skip_next_arg(arg, kind))
+		.enumerate()
+		.filter_map(|(idx, fn_arg)| match fn_arg {
+			FnArg::Receiver(_) => None,
+			FnArg::Typed(pat_type) => {
+				if let Type::Path(type_path) = &*pat_type.ty {
+					type_path
+						.path
+						.segments
+						.last()
+						.map(|segment| (idx, fn_arg, &segment.ident))
+				} else {
+					None
+				}
+			}
+		})
+}
+
+fn check_path_extractor(item_fn: &ItemFn, kind: FunctionKind) -> TokenStream {
+	let path_extractors = extractor_idents(item_fn, kind)
+		.filter(|(_, _, ident)| *ident == "Path")
+		.collect::<Vec<_>>();
+
+	if path_extractors.len() > 1 {
+		path_extractors
+			.into_iter()
+			.map(|(_, arg, _)| {
+				syn::Error::new_spanned(
+					arg,
+					"Multiple parameters must be extracted with a tuple `Path<(_, _)>` or a struct \
+					 `Path<YourParams>`, not by applying multiple `Path<_>` extractors",
+				)
+				.to_compile_error()
+			})
+			.collect()
+	} else {
+		quote! {}
+	}
+}
+
+fn is_self_pat_type(typed: &syn::PatType) -> bool {
+	let ident = if let syn::Pat::Ident(ident) = &*typed.pat {
+		&ident.ident
+	} else {
+		return false;
+	};
+
+	ident == "self"
+}
+
+fn check_inputs_impls_from_request(
+	item_fn: &ItemFn,
+	state_ty: Type,
+	kind: FunctionKind,
+) -> TokenStream {
+	let takes_self = item_fn.sig.inputs.first().map_or(false, |arg| match arg {
+		FnArg::Receiver(_) => true,
+		FnArg::Typed(typed) => is_self_pat_type(typed),
+	});
+
+	WithPosition::new(
+		item_fn
+			.sig
+			.inputs
+			.iter()
+			.filter(|arg| skip_next_arg(arg, kind)),
+	)
+	.enumerate()
+	.map(|(idx, arg)| {
+		let must_impl_from_request_parts = match &arg {
+			Position::First(_) | Position::Middle(_) => true,
+			Position::Last(_) | Position::Only(_) => false,
+		};
+
+		let arg = arg.into_inner();
+
+		let (span, ty) = match arg {
+			FnArg::Receiver(receiver) => {
+				if receiver.reference.is_some() {
+					return syn::Error::new_spanned(receiver, "Handlers must only take owned values")
+						.into_compile_error();
+				}
+
+				let span = receiver.span();
+				(span, syn::parse_quote!(Self))
+			}
+			FnArg::Typed(typed) => {
+				let ty = &typed.ty;
+				let span = ty.span();
+
+				if is_self_pat_type(typed) {
+					(span, syn::parse_quote!(Self))
+				} else {
+					(span, ty.clone())
+				}
+			}
+		};
+
+		let consumes_request = request_consuming_type_name(&ty).is_some();
+
+		let check_fn = format_ident!(
+			"__axum_macros_check_{}_{}_from_request_check",
+			item_fn.sig.ident,
+			idx,
+			span = span,
+		);
+
+		let call_check_fn = format_ident!(
+			"__axum_macros_check_{}_{}_from_request_call_check",
+			item_fn.sig.ident,
+			idx,
+			span = span,
+		);
+
+		let call_check_fn_body = if takes_self {
+			quote_spanned! {span=>
+					Self::#check_fn();
+			}
+		} else {
+			quote_spanned! {span=>
+					#check_fn();
+			}
+		};
+
+		let check_fn_generics = if must_impl_from_request_parts || consumes_request {
+			quote! {}
+		} else {
+			quote! { <M> }
+		};
+
+		let from_request_bound = if must_impl_from_request_parts {
+			quote_spanned! {span=>
+					#ty: ::axum::extract::FromRequestParts<#state_ty> + Send
+			}
+		} else if consumes_request {
+			quote_spanned! {span=>
+					#ty: ::axum::extract::FromRequest<#state_ty> + Send
+			}
+		} else {
+			quote_spanned! {span=>
+					#ty: ::axum::extract::FromRequest<#state_ty, M> + Send
+			}
+		};
+
+		quote_spanned! {span=>
+				#[allow(warnings)]
+				#[doc(hidden)]
+				fn #check_fn #check_fn_generics()
+				where
+						#from_request_bound,
+				{}
+
+				// we have to call the function to actually trigger a compile error
+				// since the function is generic, just defining it is not enough
+				#[allow(warnings)]
+				#[doc(hidden)]
+				fn #call_check_fn()
+				{
+						#call_check_fn_body
+				}
+		}
+	})
+	.collect::<TokenStream>()
+}
+
+fn check_output_tuples(item_fn: &ItemFn) -> TokenStream {
+	let elems = match &item_fn.sig.output {
+		ReturnType::Type(_, ty) => match &**ty {
+			Type::Tuple(tuple) => &tuple.elems,
+			_ => return quote! {},
+		},
+		ReturnType::Default => return quote! {},
+	};
+
+	let handler_ident = &item_fn.sig.ident;
+
+	match elems.len() {
+		0 => quote! {},
+		n if n > 17 => syn::Error::new_spanned(
+			&item_fn.sig.output,
+			"Cannot return tuples with more than 17 elements",
+		)
+		.to_compile_error(),
+		_ => WithPosition::new(elems)
+			.enumerate()
+			.map(|(idx, arg)| match arg {
+				Position::First(ty) => match extract_clean_typename(ty).as_deref() {
+					Some("StatusCode" | "Response") => quote! {},
+					Some("Parts") => check_is_response_parts(ty, handler_ident, idx),
+					Some(_) | None => {
+						if let Some(tn) = well_known_last_response_type(ty) {
+							syn::Error::new_spanned(
+								ty,
+								format!("`{tn}` must be the last element in a response tuple"),
+							)
+							.to_compile_error()
+						} else {
+							check_into_response_parts(ty, handler_ident, idx)
+						}
+					}
+				},
+				Position::Middle(ty) => {
+					if let Some(tn) = well_known_last_response_type(ty) {
+						syn::Error::new_spanned(
+							ty,
+							format!("`{tn}` must be the last element in a response tuple"),
+						)
+						.to_compile_error()
+					} else {
+						check_into_response_parts(ty, handler_ident, idx)
+					}
+				}
+				Position::Last(ty) | Position::Only(ty) => check_into_response(handler_ident, ty),
+			})
+			.collect::<TokenStream>(),
+	}
+}
+
+fn check_into_response(handler: &Ident, ty: &Type) -> TokenStream {
+	let (span, ty) = (ty.span(), ty.clone());
+
+	let check_fn = format_ident!(
+		"__axum_macros_check_{handler}_into_response_check",
+		span = span,
+	);
+
+	let call_check_fn = format_ident!(
+		"__axum_macros_check_{handler}_into_response_call_check",
+		span = span,
+	);
+
+	let call_check_fn_body = quote_spanned! {span=>
+			#check_fn();
+	};
+
+	let from_request_bound = quote_spanned! {span=>
+			#ty: ::axum_codec::response::IntoCodecResponse
+	};
+	quote_spanned! {span=>
+			#[allow(warnings)]
+			#[allow(unreachable_code)]
+			#[doc(hidden)]
+			fn #check_fn()
+			where
+					#from_request_bound,
+			{}
+
+			// we have to call the function to actually trigger a compile error
+			// since the function is generic, just defining it is not enough
+			#[allow(warnings)]
+			#[allow(unreachable_code)]
+			#[doc(hidden)]
+			fn #call_check_fn() {
+					#call_check_fn_body
+			}
+	}
+}
+
+fn check_is_response_parts(ty: &Type, ident: &Ident, index: usize) -> TokenStream {
+	let (span, ty) = (ty.span(), ty.clone());
+
+	let check_fn = format_ident!(
+		"__axum_macros_check_{}_is_response_parts_{index}_check",
+		ident,
+		span = span,
+	);
+
+	quote_spanned! {span=>
+			#[allow(warnings)]
+			#[allow(unreachable_code)]
+			#[doc(hidden)]
+			fn #check_fn(parts: #ty) -> ::axum::http::response::Parts {
+					parts
+			}
+	}
+}
+
+fn check_into_response_parts(ty: &Type, ident: &Ident, index: usize) -> TokenStream {
+	let (span, ty) = (ty.span(), ty.clone());
+
+	let check_fn = format_ident!(
+		"__axum_codec_check_{}_into_response_parts_{index}_check",
+		ident,
+		span = span,
+	);
+
+	let call_check_fn = format_ident!(
+		"__axum_codec_check_{}_into_response_parts_{index}_call_check",
+		ident,
+		span = span,
+	);
+
+	let call_check_fn_body = quote_spanned! {span=>
+			#check_fn();
+	};
+
+	let from_request_bound = quote_spanned! {span=>
+			#ty: ::axum::response::IntoResponseParts
+	};
+	quote_spanned! {span=>
+			#[allow(warnings)]
+			#[allow(unreachable_code)]
+			#[doc(hidden)]
+			fn #check_fn()
+			where
+					#from_request_bound,
+			{}
+
+			// we have to call the function to actually trigger a compile error
+			// since the function is generic, just defining it is not enough
+			#[allow(warnings)]
+			#[allow(unreachable_code)]
+			#[doc(hidden)]
+			fn #call_check_fn() {
+					#call_check_fn_body
+			}
+	}
+}
+
+fn check_input_order(item_fn: &ItemFn, kind: FunctionKind) -> Option<TokenStream> {
+	let number_of_inputs = item_fn
+		.sig
+		.inputs
+		.iter()
+		.filter(|arg| skip_next_arg(arg, kind))
+		.count();
+
+	let types_that_consume_the_request = item_fn
+		.sig
+		.inputs
+		.iter()
+		.filter(|arg| skip_next_arg(arg, kind))
+		.enumerate()
+		.filter_map(|(idx, arg)| {
+			let ty = match arg {
+				FnArg::Typed(pat_type) => &*pat_type.ty,
+				FnArg::Receiver(_) => return None,
+			};
+			let type_name = request_consuming_type_name(ty)?;
+
+			Some((idx, type_name, ty.span()))
+		})
+		.collect::<Vec<_>>();
+
+	if types_that_consume_the_request.is_empty() {
+		return None;
+	};
+
+	// exactly one type that consumes the request
+	if types_that_consume_the_request.len() == 1 {
+		// and that is not the last
+		if types_that_consume_the_request[0].0 != number_of_inputs - 1 {
+			let (_idx, type_name, span) = &types_that_consume_the_request[0];
+			let error = format!(
+				"`{type_name}` consumes the request body and thus must be the last argument to the \
+				 handler function"
+			);
+			return Some(quote_spanned! {*span=>
+					compile_error!(#error);
+			});
+		} else {
+			return None;
+		}
+	}
+
+	if types_that_consume_the_request.len() == 2 {
+		let (_, first, _) = &types_that_consume_the_request[0];
+		let (_, second, _) = &types_that_consume_the_request[1];
+		let error = format!(
+			"Can't have two extractors that consume the request body. `{first}` and `{second}` both do \
+			 that.",
+		);
+		let span = item_fn.sig.inputs.span();
+		Some(quote_spanned! {span=>
+				compile_error!(#error);
+		})
+	} else {
+		let types = WithPosition::new(types_that_consume_the_request)
+			.map(|pos| match pos {
+				Position::First((_, type_name, _)) | Position::Middle((_, type_name, _)) => {
+					format!("`{type_name}`, ")
+				}
+				Position::Last((_, type_name, _)) => format!("and `{type_name}`"),
+				Position::Only(_) => unreachable!(),
+			})
+			.collect::<String>();
+
+		let error = format!(
+			"Can't have more than one extractor that consume the request body. {types} all do that.",
+		);
+		let span = item_fn.sig.inputs.span();
+		Some(quote_spanned! {span=>
+				compile_error!(#error);
+		})
+	}
+}
+
+fn extract_clean_typename(ty: &Type) -> Option<String> {
+	let path = match ty {
+		Type::Path(type_path) => &type_path.path,
+		_ => return None,
+	};
+	path.segments.last().map(|p| p.ident.to_string())
+}
+
+fn request_consuming_type_name(ty: &Type) -> Option<&'static str> {
+	let typename = extract_clean_typename(ty)?;
+
+	let type_name = match &*typename {
+		"Json" => "Json<_>",
+		"RawBody" => "RawBody<_>",
+		"RawForm" => "RawForm",
+		"Multipart" => "Multipart",
+		"Protobuf" => "Protobuf",
+		"JsonLines" => "JsonLines<_>",
+		"Form" => "Form<_>",
+		"Request" => "Request<_>",
+		"Bytes" => "Bytes",
+		"String" => "String",
+		"Parts" => "Parts",
+		_ => return None,
+	};
+
+	Some(type_name)
+}
+
+fn well_known_last_response_type(ty: &Type) -> Option<&'static str> {
+	let typename = match extract_clean_typename(ty) {
+		Some(tn) => tn,
+		None => return None,
+	};
+
+	let type_name = match &*typename {
+		"Json" => "Json<_>",
+		"Protobuf" => "Protobuf",
+		"JsonLines" => "JsonLines<_>",
+		"Form" => "Form<_>",
+		"Bytes" => "Bytes",
+		"String" => "String",
+		_ => return None,
+	};
+
+	Some(type_name)
+}
+
+fn check_output_impls_into_response(item_fn: &ItemFn) -> TokenStream {
+	let ty = match &item_fn.sig.output {
+		syn::ReturnType::Default => return quote! {},
+		syn::ReturnType::Type(_, ty) => ty,
+	};
+	let span = ty.span();
+
+	let declare_inputs = item_fn
+		.sig
+		.inputs
+		.iter()
+		.filter_map(|arg| match arg {
+			FnArg::Receiver(_) => None,
+			FnArg::Typed(pat_ty) => {
+				let pat = &pat_ty.pat;
+				let ty = &pat_ty.ty;
+				Some(quote! {
+						let #pat: #ty = panic!();
+				})
+			}
+		})
+		.collect::<TokenStream>();
+
+	let block = &item_fn.block;
+
+	let make_value_name = format_ident!(
+		"__axum_macros_check_{}_into_response_make_value",
+		item_fn.sig.ident
+	);
+
+	let make = if item_fn.sig.asyncness.is_some() {
+		quote_spanned! {span=>
+				#[allow(warnings)]
+				#[allow(unreachable_code)]
+				#[doc(hidden)]
+				async fn #make_value_name() -> #ty {
+						#declare_inputs
+						#block
+				}
+		}
+	} else {
+		quote_spanned! {span=>
+				#[allow(warnings)]
+				#[allow(unreachable_code)]
+				#[doc(hidden)]
+				fn #make_value_name() -> #ty {
+						#declare_inputs
+						#block
+				}
+		}
+	};
+
+	let name = format_ident!("__axum_macros_check_{}_into_response", item_fn.sig.ident);
+
+	if let Some(receiver) = self_receiver(item_fn) {
+		quote_spanned! {span=>
+				#make
+
+				#[allow(warnings)]
+				#[allow(unreachable_code)]
+				#[doc(hidden)]
+				async fn #name() {
+						let value = #receiver #make_value_name().await;
+						fn check<T>(_: T)
+								where T: ::axum_codec::response::IntoCodecResponse
+						{}
+						check(value);
+				}
+		}
+	} else {
+		quote_spanned! {span=>
+				#[allow(warnings)]
+				#[allow(unreachable_code)]
+				#[doc(hidden)]
+				async fn #name() {
+						#make
+
+						let value = #make_value_name().await;
+
+						fn check<T>(_: T)
+						where T: ::axum_codec::response::IntoCodecResponse
+						{}
+
+						check(value);
+				}
+		}
+	}
+}
+
+fn check_future_send(item_fn: &ItemFn, kind: FunctionKind) -> TokenStream {
+	if item_fn.sig.asyncness.is_none() {
+		match &item_fn.sig.output {
+			syn::ReturnType::Default => {
+				return syn::Error::new_spanned(
+					item_fn.sig.fn_token,
+					format!("{} must be `async fn`s", kind.name_uppercase_plural()),
+				)
+				.into_compile_error();
+			}
+			syn::ReturnType::Type(_, ty) => ty,
+		};
+	}
+
+	let span = item_fn.sig.ident.span();
+
+	let handler_name = &item_fn.sig.ident;
+
+	let args = item_fn.sig.inputs.iter().map(|_| {
+		quote_spanned! {span=> panic!() }
+	});
+
+	let name = format_ident!("__axum_macros_check_{}_future", item_fn.sig.ident);
+
+	let do_check = quote! {
+			fn check<T>(_: T)
+					where T: ::std::future::Future + Send
+			{}
+			check(future);
+	};
+
+	if let Some(receiver) = self_receiver(item_fn) {
+		quote! {
+				#[allow(warnings)]
+				#[allow(unreachable_code)]
+				#[doc(hidden)]
+				fn #name() {
+						let future = #receiver #handler_name(#(#args),*);
+						#do_check
+				}
+		}
+	} else {
+		quote! {
+				#[allow(warnings)]
+				#[allow(unreachable_code)]
+				#[doc(hidden)]
+				fn #name() {
+						#item_fn
+						let future = #handler_name(#(#args),*);
+						#do_check
+				}
+		}
+	}
+}
+
+fn self_receiver(item_fn: &ItemFn) -> Option<TokenStream> {
+	let takes_self = item_fn.sig.inputs.iter().any(|arg| match arg {
+		FnArg::Receiver(_) => true,
+		FnArg::Typed(typed) => is_self_pat_type(typed),
+	});
+
+	if takes_self {
+		return Some(quote! { Self:: });
+	}
+
+	if let syn::ReturnType::Type(_, ty) = &item_fn.sig.output {
+		if let syn::Type::Path(path) = &**ty {
+			let segments = &path.path.segments;
+			if segments.len() == 1 {
+				if let Some(last) = segments.last() {
+					match &last.arguments {
+						syn::PathArguments::None if last.ident == "Self" => {
+							return Some(quote! { Self:: });
+						}
+						_ => {}
+					}
+				}
+			}
+		}
+	}
+
+	None
+}
+
+/// Given a signature like
+///
+/// ```skip
+/// #[debug_handler]
+/// async fn handler(
+///     _: axum::extract::State<AppState>,
+///     _: State<AppState>,
+/// ) {}
+/// ```
+///
+/// This will extract `AppState`.
+///
+/// Returns `None` if there are no `State` args or multiple of different types.
+fn state_types_from_args(item_fn: &ItemFn) -> HashSet<Type> {
+	let types = item_fn
+		.sig
+		.inputs
+		.iter()
+		.filter_map(|input| match input {
+			FnArg::Receiver(_) => None,
+			FnArg::Typed(pat_type) => Some(pat_type),
+		})
+		.map(|pat_type| &*pat_type.ty);
+	crate::infer_state_types(types).collect()
+}
+
+fn next_is_last_input(item_fn: &ItemFn) -> TokenStream {
+	let next_args = item_fn
+		.sig
+		.inputs
+		.iter()
+		.enumerate()
+		.filter(|(_, arg)| !skip_next_arg(arg, FunctionKind::Middleware))
+		.collect::<Vec<_>>();
+
+	if next_args.is_empty() {
+		return quote! {
+				compile_error!(
+						"Middleware functions must take `axum::middleware::Next` as the last argument",
+				);
+		};
+	}
+
+	if next_args.len() == 1 {
+		let (idx, arg) = &next_args[0];
+		if *idx != item_fn.sig.inputs.len() - 1 {
+			return quote_spanned! {arg.span()=>
+					compile_error!("`axum::middleware::Next` must the last argument");
+			};
+		}
+	}
+
+	if next_args.len() >= 2 {
+		return quote! {
+				compile_error!(
+						"Middleware functions can only take one argument of type `axum::middleware::Next`",
+				);
+		};
+	}
+
+	quote! {}
+}
+
+fn skip_next_arg(arg: &FnArg, kind: FunctionKind) -> bool {
+	match kind {
+		FunctionKind::Handler => true,
+		FunctionKind::Middleware => match arg {
+			FnArg::Receiver(_) => true,
+			FnArg::Typed(pat_type) => {
+				if let Type::Path(type_path) = &*pat_type.ty {
+					type_path
+						.path
+						.segments
+						.last()
+						.map_or(true, |path_segment| path_segment.ident != "Next")
+				} else {
+					true
+				}
+			}
+		},
+	}
+}
+
+#[test]
+fn ui_debug_handler() {
+	crate::run_ui_tests("debug_handler");
+}
+
+#[test]
+fn ui_debug_middleware() {
+	crate::run_ui_tests("debug_middleware");
+}

--- a/macros/src/lib.rs
+++ b/macros/src/lib.rs
@@ -1,12 +1,14 @@
+use proc_macro::TokenStream;
+use quote::{quote, ToTokens};
+use syn::parse::Parse;
+
+mod apply;
+mod attr_parsing;
+mod debug_handler;
+mod with_position;
+
 /// A utility macro for automatically deriving the correct traits
 /// depending on the enabled features.
-#[cfg(any(
-	feature = "bincode",
-	feature = "bitcode",
-	feature = "serde",
-	feature = "aide",
-	feature = "validator"
-))]
 #[proc_macro_attribute]
 pub fn apply(
 	attr: proc_macro::TokenStream,
@@ -15,227 +17,112 @@ pub fn apply(
 	apply::apply(attr, input)
 }
 
-#[cfg(any(
-	feature = "bincode",
-	feature = "bitcode",
-	feature = "serde",
-	feature = "aide",
-	feature = "validator"
-))]
-mod apply {
-	use proc_macro2::TokenStream;
-	use quote::{quote, ToTokens};
-	use syn::{
-		parse::{Parse, ParseStream},
-		punctuated::Punctuated,
-		spanned::Spanned,
-		Meta, Path, Token,
-	};
+/// Generates better error messages when applied to handler functions.
+///
+/// For more information, see [`axum::debug_handler`](https://docs.rs/axum/latest/axum/attr.debug_handler.html).
+#[proc_macro_attribute]
+pub fn debug_handler(_attr: TokenStream, input: TokenStream) -> TokenStream {
+	#[cfg(not(debug_assertions))]
+	return input;
 
-	struct Args {
-		encode: bool,
-		decode: bool,
-		crate_name: Path,
+	#[cfg(debug_assertions)]
+	return expand_attr_with(_attr, input, |attrs, item_fn| {
+		debug_handler::expand(attrs, item_fn, debug_handler::FunctionKind::Handler)
+	});
+}
+
+/// Generates better error messages when applied to middleware functions.
+///
+/// For more information, see [`axum::debug_middleware`](https://docs.rs/axum/latest/axum/attr.debug_middleware.html).
+#[proc_macro_attribute]
+pub fn debug_middleware(_attr: TokenStream, input: TokenStream) -> TokenStream {
+	#[cfg(not(debug_assertions))]
+	return input;
+
+	#[cfg(debug_assertions)]
+	return expand_attr_with(_attr, input, |attrs, item_fn| {
+		debug_handler::expand(attrs, item_fn, debug_handler::FunctionKind::Middleware)
+	});
+}
+
+fn expand_attr_with<F, A, I, K>(attr: TokenStream, input: TokenStream, f: F) -> TokenStream
+where
+	F: FnOnce(A, I) -> K,
+	A: Parse,
+	I: Parse,
+	K: ToTokens,
+{
+	let expand_result = (|| {
+		let attr = syn::parse(attr)?;
+		let input = syn::parse(input)?;
+		Ok(f(attr, input))
+	})();
+	expand(expand_result)
+}
+
+fn expand<T>(result: syn::Result<T>) -> TokenStream
+where
+	T: ToTokens,
+{
+	match result {
+		Ok(tokens) => {
+			let tokens = (quote! { #tokens }).into();
+			if std::env::var_os("AXUM_MACROS_DEBUG").is_some() {
+				eprintln!("{tokens}");
+			}
+			tokens
+		}
+		Err(err) => err.into_compile_error().into(),
 	}
+}
 
-	impl Parse for Args {
-		fn parse(input: ParseStream) -> syn::Result<Self> {
-			let options = Punctuated::<Meta, Token![,]>::parse_terminated(input)?;
-
-			let mut encode = false;
-			let mut decode = false;
-			let mut crate_name = syn::parse_str("axum_codec").expect("failed to parse crate name");
-
-			for meta in options {
-				match meta {
-					Meta::List(list) => {
-						return Err(syn::Error::new(
-							list.span(),
-							"expected `encode`, `decode`, or `crate`",
-						))
-					}
-					Meta::Path(path) => {
-						let ident = path.get_ident().map(|ident| ident.to_string());
-						match ident.as_deref() {
-							Some("encode") if encode => {
-								return Err(syn::Error::new(
-									path.span(),
-									"option `encode` is already enabled",
-								))
-							}
-							Some("decode") if decode => {
-								return Err(syn::Error::new(
-									path.span(),
-									"option `decode` is already enabled",
-								))
-							}
-							Some("encode") => encode = true,
-							Some("decode") => decode = true,
-							Some(other) => {
-								return Err(syn::Error::new(
-									path.span(),
-									format!("unknown option `{other}`, expected `encode` or `decode`"),
-								))
-							}
-							None => {
-								return Err(syn::Error::new(
-									path.span(),
-									"expected `encode` or `decode`",
-								))
-							}
-						}
-					}
-					Meta::NameValue(name_value) => {
-						if !name_value.path.is_ident("crate") {
-							return Err(syn::Error::new(name_value.path.span(), "expected `crate`"));
-						}
-
-						let path = match name_value.value {
-							syn::Expr::Lit(ref lit) => match &lit.lit {
-								syn::Lit::Str(path) => path,
-								_ => return Err(syn::Error::new(lit.span(), "expected a string")),
-							},
-							_ => {
-								return Err(syn::Error::new(
-									name_value.value.span(),
-									"expected a literal string",
-								))
-							}
-						};
-
-						let mut path = syn::parse_str::<Path>(&path.value()).expect("failed to parse path");
-
-						path.leading_colon = if path.is_ident("crate") {
-							None
-						} else {
-							Some(Token![::](name_value.value.span()))
-						};
-
-						crate_name = path;
-					}
+fn infer_state_types<'a, I>(types: I) -> impl Iterator<Item = syn::Type> + 'a
+where
+	I: Iterator<Item = &'a syn::Type> + 'a,
+{
+	types
+		.filter_map(|ty| {
+			if let syn::Type::Path(path) = ty {
+				Some(&path.path)
+			} else {
+				None
+			}
+		})
+		.filter_map(|path| {
+			if let Some(last_segment) = path.segments.last() {
+				if last_segment.ident != "State" {
+					return None;
 				}
+
+				match &last_segment.arguments {
+					syn::PathArguments::AngleBracketed(args) if args.args.len() == 1 => {
+						Some(args.args.first().unwrap())
+					}
+					_ => None,
+				}
+			} else {
+				None
 			}
-
-			if !encode && !decode {
-				return Err(syn::Error::new(
-					input.span(),
-					"at least one of `encode` or `decode` must be enabled",
-				));
+		})
+		.filter_map(|generic_arg| {
+			if let syn::GenericArgument::Type(ty) = generic_arg {
+				Some(ty)
+			} else {
+				None
 			}
-
-			Ok(Self {
-				encode,
-				decode,
-				crate_name,
-			})
-		}
-	}
-
-	pub fn apply(
-		attr: proc_macro::TokenStream,
-		input: proc_macro::TokenStream,
-	) -> proc_macro::TokenStream {
-		let args = syn::parse_macro_input!(attr as Args);
-
-		let crate_name = &args.crate_name;
-		let mut tokens = TokenStream::default();
-
-		#[cfg(feature = "serde")]
-		{
-			if args.encode {
-				tokens.extend(quote! {
-					#[derive(#crate_name::__private::serde::Serialize)]
-				});
-			}
-
-			if args.decode {
-				tokens.extend(quote! {
-					#[derive(#crate_name::__private::serde::Deserialize)]
-				});
-			}
-
-			let crate_ = format!("{}::__private::serde", crate_name.to_token_stream());
-
-			tokens.extend(quote! {
-				#[serde(crate = #crate_)]
-			});
-		}
-
-		#[cfg(feature = "bincode")]
-		{
-			if args.encode {
-				tokens.extend(quote! {
-					#[derive(#crate_name::__private::bincode::Encode)]
-				});
-			}
-
-			if args.decode {
-				tokens.extend(quote! {
-					#[derive(#crate_name::__private::bincode::Decode)]
-				});
-			}
-
-			let crate_ = format!("{}::__private::bincode", crate_name.to_token_stream());
-
-			tokens.extend(quote! {
-				#[bincode(crate = #crate_)]
-			});
-		}
-
-		#[cfg(feature = "bitcode")]
-		{
-			if args.encode {
-				tokens.extend(quote! {
-					#[derive(#crate_name::__private::bitcode::Encode)]
-				});
-			}
-
-			if args.decode {
-				tokens.extend(quote! {
-					#[derive(#crate_name::__private::bitcode::Decode)]
-				});
-			}
-
-			let crate_ = format!("{}::__private::bitcode", crate_name.to_token_stream());
-
-			tokens.extend(quote! {
-				#[bitcode(crate = #crate_)]
-			});
-		}
-
-		#[cfg(feature = "aide")]
-		{
-			let crate_ = format!("{}::__private::schemars", crate_name.to_token_stream());
-
-			tokens.extend(quote! {
-				#[derive(#crate_name::__private::schemars::JsonSchema)]
-				#[schemars(crate = #crate_)]
-			});
-		}
-
-		// TODO: Implement #[validate(crate = "...")]
-		// For now, use the real crate name so the error is nicer.
-		#[cfg(feature = "validator")]
-		if args.decode {
-			tokens.extend(quote! {
-				#[derive(validator::Validate)]
-			});
-		}
-
-		tokens.extend(TokenStream::from(input));
-		tokens.into()
-	}
+		})
+		.cloned()
 }
 
 #[doc(hidden)]
 #[proc_macro]
-pub fn __private_decode_trait(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+pub fn __private_decode_trait(input: TokenStream) -> TokenStream {
 	__private::decode_trait(input.into()).into()
 }
 
 #[doc(hidden)]
 #[proc_macro]
-pub fn __private_encode_trait(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+pub fn __private_encode_trait(input: TokenStream) -> TokenStream {
 	__private::encode_trait(input.into()).into()
 }
 

--- a/macros/src/with_position.rs
+++ b/macros/src/with_position.rs
@@ -1,0 +1,116 @@
+// this is copied from itertools under the following license
+//
+// Copyright (c) 2015
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::iter::{Fuse, FusedIterator, Peekable};
+
+pub(crate) struct WithPosition<I>
+where
+	I: Iterator,
+{
+	handled_first: bool,
+	peekable: Peekable<Fuse<I>>,
+}
+
+impl<I> WithPosition<I>
+where
+	I: Iterator,
+{
+	pub(crate) fn new(iter: impl IntoIterator<IntoIter = I>) -> WithPosition<I> {
+		WithPosition {
+			handled_first: false,
+			peekable: iter.into_iter().fuse().peekable(),
+		}
+	}
+}
+
+impl<I> Clone for WithPosition<I>
+where
+	I: Clone + Iterator,
+	I::Item: Clone,
+{
+	fn clone(&self) -> Self {
+		Self {
+			handled_first: self.handled_first,
+			peekable: self.peekable.clone(),
+		}
+	}
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub(crate) enum Position<T> {
+	First(T),
+	Middle(T),
+	Last(T),
+	Only(T),
+}
+
+impl<T> Position<T> {
+	pub(crate) fn into_inner(self) -> T {
+		match self {
+			Position::First(x) | Position::Middle(x) | Position::Last(x) | Position::Only(x) => x,
+		}
+	}
+}
+
+impl<I: Iterator> Iterator for WithPosition<I> {
+	type Item = Position<I::Item>;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		match self.peekable.next() {
+			Some(item) => {
+				if !self.handled_first {
+					// Haven't seen the first item yet, and there is one to give.
+					self.handled_first = true;
+					// Peek to see if this is also the last item,
+					// in which case tag it as `Only`.
+					match self.peekable.peek() {
+						Some(_) => Some(Position::First(item)),
+						None => Some(Position::Only(item)),
+					}
+				} else {
+					// Have seen the first item, and there's something left.
+					// Peek to see if this is the last item.
+					match self.peekable.peek() {
+						Some(_) => Some(Position::Middle(item)),
+						None => Some(Position::Last(item)),
+					}
+				}
+			}
+			// Iterator is finished.
+			None => None,
+		}
+	}
+
+	fn size_hint(&self) -> (usize, Option<usize>) {
+		self.peekable.size_hint()
+	}
+}
+
+impl<I> ExactSizeIterator for WithPosition<I> where I: ExactSizeIterator {}
+
+impl<I: Iterator> FusedIterator for WithPosition<I> {}

--- a/src/handler.rs
+++ b/src/handler.rs
@@ -20,7 +20,7 @@ impl<T> Input for T where T: aide::OperationInput {}
 
 #[diagnostic::on_unimplemented(
 	note = "Consider wrapping the return value in `Codec` if appropriate",
-	note = "Consider using `#[axum::debug_handler]` to improve the error message"
+	note = "Consider using `#[axum_codec::debug_handler]` to improve the error message"
 )]
 pub trait CodecHandler<T, I: Input, D, S>: Clone + Send + 'static {
 	type Future: Future<Output = Response> + Send;

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -49,6 +49,10 @@ pub mod __private {
 pub use axum_codec_macros as macros;
 #[cfg(feature = "macros")]
 pub use macros::apply;
+#[cfg(feature = "macros")]
+pub use macros::debug_handler;
+#[cfg(feature = "macros")]
+pub use macros::debug_middleware;
 
 #[cfg(test)]
 mod test {

